### PR TITLE
Fix fastlane so it signs and submit builds

### LINKED ISSUE #3205: Fix build (and particularly signing) with Fastlane
I was trying to do this a while back, but for some reason, signing just doesn't work for me.

PS: try not to upload. You should be able to fix it without uploading.

### PR DESCRIPTION
**Pull request by eviltofu**
Fixes #3205 

I have a possible fix. Can @oa-s and @hboon test to see if this works? 

```
make clean && make install_bundle && make bootstrap && make release
```

**Note (1-old):**

I reduced the `Fastfile` to the following lines below and I was able to proceed until the upload to App Store stage (which I cut short by hitting control-c). I then retrofitted the line back into the original `Fastfile`.

```ruby
default_platform(:ios)

platform :ios do
  desc "Push a new release build to the App Store"
  lane :release do
    increment_build_number(xcodeproj: "AlphaWallet.xcodeproj")
    build_app(workspace: "AlphaWallet.xcworkspace", scheme: "AlphaWallet", xcargs: "-allowProvisioningUpdates")
    upload_to_app_store(skip_metadata: true, skip_screenshots: true, skip_binary_upload: true)
  end
end
```

**Note (2-version numbers):**

Somehow the `fastlane` script was expecting strings in "x.y.z" in the Info.plist files but the Info.plist files had expressions `($CURRENT_PROJECT)` etc which was confusing it. The version number has to be in the "x.y.z" form or the `fastlane` script just barfs.

### CODE REVIEW

**Comment by hboon** [comment]
@eviltofu sure. I'll try it out. Thanks!

**Comment by eviltofu** [comment]
> @eviltofu sure. I'll try it out. Thanks!

I switched out the 
```ruby
    cert
    match(type: "appstore", readonly: true) 
```
in my fix but did not in the original file.

**Comment by hboon** [comment]
@eviltofu I don't quite understand, but can you help amend the commit (if necessary) with the version you think will work? I'll test it out. Probably not today.

**Comment by hboon** [comment]
@eviltofu Can you describe how you ran it?

I commented out:

```
#increment_build_number
#ensure_git_branch()
#increment_version_number(bump_type: "patch")
```

2. Then:

```
./vendor/bundle/bin/bundle exec fastlane^C
```
3. And chose 1 (for `ios release`)

4. And it shows:

```
[12:31:17]: Cloning remote git repo...
[12:31:17]: If cloning the repo takes too long, you can use the `clone_branch_directly` option in match.
Cloning into '/var/folders/7w/yw8_hz1d6mxgvdt0j_5dxxb80000gq/T/d20220203-23522-kskyjr'...
trustwallet@bitbucket.org: Permission denied (publickey).
fatal: Could not read from remote repository.
Please make sure you have the correct access rights and the repository exists.
[12:31:20]: Exit status: 128
[12:31:20]: Error cloning certificates repo, please make sure you have read access to the repository you want to use
[12:31:20]: Run the following command manually to make sure you're properly authenticated:
[12:31:20]: $ git clone trustwallet@bitbucket.org:trustwallet/certificates.git /var/folders/7w/yw8_hz1d6mxgvdt0j_5dxxb80000gq/T/d2022020
3-23522-kskyjr

**Comment by hboon** [comment]
Also would you help change the commit message to something like:

"Fix fastlane so it signs and submit builds".

This is less useful in the logs:

"Replaced Gym with build_app in Fastfile #3855"

**Comment by eviltofu** [comment]
Don’t try this. Was working on it this morning and found a bug.

**Comment by eviltofu** [comment]
@hboon I updated the instructions in the first message on how to run this. Do not merge as I created a new lane called `testrelease` and ran this from the `Makefile`. Once this is acceptable, I will rename the lane to release and update the make files.

**Comment by hboon** [comment]
@eviltofu ready to test?

**Comment by eviltofu** [comment]
> @eviltofu ready to test?

@hboon follow instructions at top.

File: AlphaWallet.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -5117,7 +5117,7 @@
 			buildPhases = (
 				7C3903F7FBE1023CAA5AED0C /* [CP] Check Pods Manifest.lock */,
 				87D5BBC12727CBB70053E6D2 /* Apollo */,
-				C83F787024A1C1DB00783B67 /* Check correct credentials used for API keys */,
+				028B36D327ABCFF60054F172 /* Check correct credentials used for API keys */,
```
Comment (hboon) [inline AlphaWallet.xcodeproj/project.pbxproj:5120]: Can you try to exclude this?

File: AlphaWallet.xcodeproj/project.pbxproj [hunk 2]
```diff
@@ -5356,6 +5356,24 @@
 /* End PBXResourcesBuildPhase section */
 
 /* Begin PBXShellScriptBuildPhase section */
+		028B36D327ABCFF60054F172 /* Check correct credentials used for API keys */ = {
```
Comment (hboon) [inline AlphaWallet.xcodeproj/project.pbxproj:5359]: Can you avoid changing this?

File: AlphaWallet.xcodeproj/project.pbxproj [hunk 3]
```diff
@@ -6029,7 +6029,7 @@
 				87CC3853273E7980005BE9B8 /* eip155URLDecoder.swift in Sources */,
 				5E7C7ECE164289A89734B4EF /* LocalesCoordinator.swift in Sources */,
 				874D2F8027A2B1D1005459DA /* UnstoppableDomainsV2Resolver.swift in Sources */,
-				879F184826E73CF4000602F2 /* Erc1155TokenInstanceViewController.swift in Sources */,
```
Comment (hboon) [inline AlphaWallet.xcodeproj/project.pbxproj:6032]: Looks like the Xcode project file was broken, I have created a separate PR #3872. Can you remove this later?

File: AlphaWallet.xcodeproj/project.pbxproj [hunk 4]
```diff
@@ -6833,7 +6833,7 @@
 			isa = XCBuildConfiguration;
 			baseConfigurationReference = CA4F65A76A018D8DFF4A0314 /* Pods-AlphaWallet.debug.xcconfig */;
 			buildSettings = {
-				ALPHAWALLET_INFO_PLIST_FILE = AlphaWallet/Info.plist;
```
Comment (hboon) [inline AlphaWallet.xcodeproj/project.pbxproj:6836]: Does the original reference not work?

**Review by hboon — COMMENTED** [review]
Got this error messages:

> [12:14:29]: ▸ error: No profiles for 'com.stormbird.alphawallet.AlphaWalletShare' were found: Xcode couldn't find any iOS App Development provisioning profiles matching 'com.stormbird.alphawallet.AlphaWalletShare'. (in target 'AlphaWalletShare' from project 'AlphaWallet')
> [12:14:29]: ▸ error: No profiles for 'com.stormbird.alphawallet' were found: Xcode couldn't find any iOS App Development provisioning profiles matching 'com.stormbird.alphawallet'. (in target 'AlphaWallet' from project 'AlphaWallet')

Any idea why?

**Comment by hboon** [comment]
@eviltofu any luck with this?

**Comment by eviltofu** [comment]
> @eviltofu any luck with this?

No. I don't get that error. I'm wondering how you have the signing permissions set up.

**Comment by hboon** [comment]
> No. I don't get that error. I'm wondering how you have the signing permissions set up.

One problem I have is automatic signing in Xcode has never worked for me as far as I can remember since the first version of iPhone OS SDK. Related?

**Comment by eviltofu** [comment]
@hboon How does one test the uploading to the App Store `deliver(force: true)`? I've commented out this action in the `Fastfile`. 

I've pushed another update for you to test. I've removed all the build phases/checks which stop this from building. I will push another update to re-enable them when this passes.

Try `make clean && make install_bundle && make bootstrap && make release`. If that fails, try the following:
`FASTLANE_USER=xxx@yyy.zzz FASTLANE_PASSWORD=abcdefg make clean && make install_bundle && make bootstrap && make release` where `FASTLANE_USER` and `FASTLANE_PASSWORD` are your Apple ID and password. If that fails, post the error messages here and I'll try to debug further.

**Comment by hboon** [comment]
@eviltofu we can create another app ID or we can use our own accounts to test. Have you tried with your own account and if it works for you? Probably best way is yo create a separate app ID then we can replicate with the same cert and team ID. Any preference?

**Comment by eviltofu** [comment]
> @eviltofu we can create another app ID or we can use our own accounts to test. Have you tried with your own account and if it works for you? Probably best way is yo create a separate app ID then we can replicate with the same cert and team ID. Any preference?

@hboon I'll make another app ID on my account and continue testing it there. Can you verify that the current script allows you to sign at least? Including the `FASTLANE_USER` work around?

**Comment by hboon** [comment]
No, it complains about profiles:

> error: No profiles for 'com.stormbird.alphawallet' were found: Xcode couldn't find any iOS App Development provisioning profiles matching 'com.stormbird.alphawallet'. (in target 'AlphaWallet' from project 'AlphaWallet')
> error: No profiles for 'com.stormbird.alphawallet.AlphaWalletSafariExtension' were found: Xcode couldn't find any iOS App Development provisioning profiles matching 'com.stormbird.alphawallet.AlphaWalletSafariExtension'. (in target 'AlphaWalletSafariExtension' from project 'AlphaWallet')
> error: No profiles for 'com.stormbird.alphawallet.AlphaWalletShare' were found: Xcode couldn't find any iOS App Development provisioning profiles matching 'com.stormbird.alphawallet.AlphaWalletShare'. (in target 'AlphaWalletShare' from project 'AlphaWallet')

This is both with and without `FASTLANE_USER` and `FASTLANE_PASSWORD`. Leaving it out doesn't complain about invalid user/password too. Earlier versions of this PR seems to complain when it's omitted (I remember I had to modify one of the files where it was hardcoded)

**Comment by eviltofu** [comment]
@hboon can you show me this screen in Xcode?

Screenshot 2022-03-25 at 11 44 36 AM

**Comment by hboon** [comment]
Screenshot 2022-03-25 at 11 54 20 AM

Automatic signing has never worked for me as far as I can remember though. Maybe something wrong with my environment?

**Comment by eviltofu** [comment]
@hboon how do you release to TestFlight etc?

**Comment by hboon** [comment]
Build like normal and manually select certs and profiles

**Comment by eviltofu** [comment]
Searching online for `Xcode couldn't find any iOS App Development provisioning profiles matching` generates the following suggestions:

1. You should login to your developer account and accept the new agreements etc.
2. Logout and login from your developer account in Xcode then download your certificates.
3. Use manual signing. [Link](https://troubleshootpedia.com/xcode-couldnt-find-any-ios-app-development-provisioning-profiles-matching/).
4. Delete everything in `~/Library/MobileDevice/Provisioning Profiles/`. [Link](https://4gbritain.org/xcode-couldnt-find-any-ios-app-development-provisioning-profiles-matching/).

@hboon

**Comment by hboon** [comment]
> You should login to your developer account and accept the new agreements etc.

Verified.

> Logout and login from your developer account in Xcode then download your certificates.

Done.

> Use manual signing. Link.

Have always worked. Meaning certificate is OK

> Delete everything in ~/Library/MobileDevice/Provisioning Profiles/. Link.

I removed all the provisioning profiles and then use Xcode to sign and upload and previously archived build. This is what I see if I choose automatic:

Screenshot 2022-03-28 at 11 46 13 AM

I noticed that ~/Library/MobileDevice/Provisioning Profiles/ seems to contain downloaded profiles:

Screenshot 2022-03-28 at 11 51 46 AM

If I change it to manual signing, I see this (notice the distribution cert being available):

Screenshot 2022-03-28 at 11 46 29 AM

**Comment by eviltofu** [comment]
@hboon What does the [identifier link](https://developer.apple.com/account/resources/identifiers/list) at apple.com show?

**Comment by eviltofu** [comment]
Also can @oa-s test this to check if @hboon or I am the outlier?

**Comment by hboon** [comment]
> @hboon What does the [identifier link](https://developer.apple.com/account/resources/identifiers/list) at apple.com show?

```
Stormbird Pte. Ltd. - LRAW5PL536
```

Which is matches the project setting

**Comment by eviltofu** [comment]
@hboon https://developer.apple.com/account/#!/membership/LRAW5PL536 What roles do you have? Is **Developer** one o them?

**Comment by hboon** [comment]
> What roles do you have

I'm the account admin.

Relevant part from `cert` is:

```
More information about Two-factor Authentication: https://support.apple.com/en-us/HT204915
If you're running this in a non-interactive session (e.g. server or CI)
check out https://github.com/fastlane/fastlane/tree/master/spaceship#2-step-verification

Please enter the 6 digit code you received at +65 •••• ••96:
482788
Requesting session...                                                                                                                  [22:24:40]: Successfully logged in
[22:24:41]: Found the certificate RLX67HH29J (Stormbird Pte. Ltd.) which is installed on the local machine. Using this one.
[22:24:42]: Verifying the certificate is properly installed locally...
[22:24:42]: Successfully installed certificate RLX67HH29J
[22:24:42]: Use signing certificate 'RLX67HH29J' from now on!
[22:24:42]: Result: RLX67HH29J

```
Relevant part from `sigh` is:
```
[22:25:29]: Successfully logged in
[22:25:29]: Fetching profiles...
[22:25:29]: Verifying certificates...
[22:25:30]: Found 2 matching profile(s)
[22:25:30]: Downloading provisioning profile...
[22:25:30]: Successfully downloaded provisioning profile...
[22:25:30]: Installing provisioning profile...
/Users/hboon/Documents/workspace/projects/alphawallet/alpha-wallet-ios/AppStore_com.stormbird.alphawallet.mobileprovision
[22:25:30]: Setting Provisioning Profile type to 'app-store'
[22:25:30]: Result: d7eab76c-2c01-4dcb-8629-9a11b893e369
```

And then, the cert appears at root of repo:

```
RLX67HH29J.cer
```

**Comment by hboon** [comment]
@eviltofu closed?

**Pull request closed without merging**